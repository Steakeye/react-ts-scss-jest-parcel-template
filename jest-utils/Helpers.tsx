import React, { ReactElement } from 'react';
import { BrowserRouter, RouteComponentProps } from 'react-router-dom';
import { renderToString } from 'react-dom/server';
import {
  createMemoryHistory,
  MemoryHistoryBuildOptions,
  Location,
} from 'history';
import { match } from 'react-router';
import { RenderResult } from '@testing-library/react';
// tslint:disable-next-line:ban-ts-ignore
// @ts-ignore
import reactElementToJSXString from 'react-element-to-jsx-string';
import { Optional, POJSObject } from '~/app/declarations/standard';

export interface RouteWrapperConfig {
  children?: JSX.Element[];
}

export function routerWrapper({ children }: RouteWrapperConfig) {
  return <BrowserRouter>{children}</BrowserRouter>;
}

export function svgWrapper(children: JSX.Element | JSX.Element[]) {
  return <svg>{children}</svg>;
}

export function getElementByQuery(
  context: RenderResult | Element,
  query: string
): Optional<Element> {
  const el = (context as RenderResult).container
    ? (context as RenderResult).container
    : (context as Element);

  return el.querySelector(query);
}

function mutateProps(props: POJSObject, stringifyObjects = true): void {
  let key: string;
  // tslint:disable-next-line:forin
  for (key in props) {
    const val = props[key];
    if (key !== 'className') {
      delete props[key];
      key = key.toLowerCase();
    }
    if (key !== 'disabled' && typeof val === 'boolean') {
      props[key] = val.toString();
    } else if (
      val &&
      typeof (val as { $$typeof: unknown }).$$typeof === 'symbol'
    ) {
      let parsedVal: string;

      if (typeof (val as { type: unknown }).type === 'string') {
        parsedVal = reactElementToJSXString(val);
      } else {
        parsedVal = renderToString(val as ReactElement);
      }

      props[key] = parsedVal;
    } else if (stringifyObjects && typeof val === 'object') {
      props[key] = JSON.stringify(val as {});
    } else {
      props[key] = val;
    }
  }
}

export interface MockRouteComponentProps<P> {
  history?: Partial<MemoryHistoryBuildOptions>;
  location?: Partial<Location>;
  match?: Partial<match<P>>;
}

export function createRouteComponentProps<T>({
  history,
  location,
  match,
}: MockRouteComponentProps<T> = {}): RouteComponentProps<T> {
  const defaultLocation = {
    pathname: '',
    search: '',
    state: '',
    hash: '',
  };
  const defaultMatch = { params: {} as T, isExact: true, path: '', url: '' };

  return {
    history: createMemoryHistory(history),
    location: { ...defaultLocation, ...location },
    match: { ...defaultMatch, ...match },
  };
}

export function createMockComponent(
  componentName: string,
  testId: string,
  stringifyObjects = true
) {
  return jest.fn(props => {
    const { children, ...restOfProps } = props;

    mutateProps(restOfProps, stringifyObjects);

    return (
      <span data-mock={componentName} data-testid={testId} {...restOfProps}>
        {children ? children : `Mock ${componentName}`}
      </span>
    );
  });
}

export function createMockContextConsumer(
  componentName: string,
  testId: string,
  mockState: {} = {}
): jest.Mock<React.Consumer<{}>> {
  return jest.fn(props => {
    const { children } = props;

    const childrenResult = children(mockState);

    return childrenResult;
  });
}

export function getMockCallValue<T extends unknown>(
  mock: jest.Mock,
  index = 0
): T[] {
  return mock.mock.calls[index] as T[];
}

export function getMockReturnValue<T extends unknown>(
  mock: jest.Mock,
  index = 0
): T {
  return mock.mock.results[index].value as T;
}

export function createMockAppEnums() {
  enum mockBaseReducerActionType {
    Create = 'mock-create',
    Delete = 'mock-delete',
    Read = 'mock-read',
    Update = 'mock-update',
  }

  enum mockAsyncReducerActionType {
    Request = 'mock-request',
    Receive = 'mock-receive',
    Request = 'mock-request',
  }

  enum mockEditableReducerActionType {
    Remove = 'mock-remove',
    Reset = 'mock-reset',
  }

  const appModule = {
    BaseReducerActionType: mockBaseReducerActionType,
    AsyncReducerActionType: mockAsyncReducerActionType,
    EditableReducerActionType: mockEditableReducerActionType,
  };

  return appModule;
}
