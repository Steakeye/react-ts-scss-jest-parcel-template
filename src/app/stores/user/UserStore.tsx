import * as React from 'react';
import { UserContext as User, useUserDataFetch } from './UserContext';

function UserStore(
  props: React.PropsWithChildren<{ children: React.ReactNode }>
) {
  const { userData } = useUserDataFetch();

  return <User.Provider value={userData}>{props.children}</User.Provider>;
}

export { UserStore };
