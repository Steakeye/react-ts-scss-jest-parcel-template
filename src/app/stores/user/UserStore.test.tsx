import * as React from 'react';
import { render, RenderResult, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createMockComponent } from '~/../jest-utils/Helpers';
import { UserStoreFC } from './UserStore.d';

const mockUser = 'mock-user';
const useUserDataFetch = jest.fn(() => ({ userData: mockUser }));

jest.doMock('./UserContext', () => ({
  UserContext: {
    Provider: createMockComponent(
      'UserStoreProvider',
      'mock-user-store-provider'
    ),
  },
  useUserDataFetch,
}));

describe('UserStore module - exposes persistent user data', () => {
  describe('UserStore component - composes the user data provider and children components', () => {
    let UserStore: UserStoreFC;
    let storesInstance: RenderResult;

    describe('expected behaviour is', () => {
      beforeAll(async () => {
        ({ UserStore } = await import('./UserStore'));
      });

      beforeEach(() => {
        act(() => {
          storesInstance = render(<UserStore>{'store children'}</UserStore>);
        });
      });

      afterEach(() => {
        if (storesInstance) {
          const { unmount } = storesInstance;
          unmount();
        }
      });

      it('user data is fetched', () => {
        expect(useUserDataFetch).toHaveBeenCalled();
      });

      it('composes the store and children', () => {
        expect(storesInstance.container).toMatchSnapshot();
        expect(storesInstance.container).toHaveTextContent('store children');
      });

      // tslint:disable-next-line:ban
      it('has data from an initial API call', () => {
        const mockStoreProvider = storesInstance.getByTestId(
          'mock-user-store-provider'
        );
        expect(mockStoreProvider).toHaveAttribute('value', mockUser);
      });
    });
  });
});
