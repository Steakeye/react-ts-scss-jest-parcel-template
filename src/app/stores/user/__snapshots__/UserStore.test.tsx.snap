// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserStore module - exposes persistent user data UserStore component - composes the user data provider and children components expected behaviour is composes the store and children 1`] = `
<div>
  <span
    data-mock="UserStoreProvider"
    data-testid="mock-user-store-provider"
    value="mock-user"
  >
    store children
  </span>
</div>
`;
